Cleaning up Controller::handleRequest

dhensby opened the pull request:
This PR reduced the amount of responsiblity taken on by `Controller::handleRequest`. In turn this reduces the code complexity as well.

The handleRequest function has been split into component parts (setup and teardown - the functions that prepare and cleanup the controller before and after the request is _actually_ handled).

The init function has had a similar treatment. The init function is now split into the logic that should actually be happening when init is run and the logic preparing to run the init and enforcing that base init has been run at all.

init is still defined on controllers and executed as expected but instantiators of controllers should be calling `doInit` from now on when preparing their controllers.

Further, I've removed all direct manipulations of class properties and used their accessors where appropriate (and created them where missing).

---

1. Separated responsibility of handleAction so that it no longer bootstraps the controller and cleans up after the request is handled.
2. NEW beforeHandleRequest to take responsibility of bootstrapping the controller
3. NEW afterHandleRequest to take responsibility of cleanup for the controller
4. NEW calling init on controllers removed in favour of doInit() which takes responsibility of enforcing that "base init" is called and the before and after hooks. init() is now protected visibility
5. NEW Added prepareResponse to Controller for dealing with responses from controllers
6. NEW setResponse added to controller for setting response objects on the controller

tractorcow commented on control/RequestHandler.php:
```diff
@@ -495,5 +495,6 @@ public function getRequest() {
 	 */
```
`@return $this`

tractorcow commented on control/Controller.php:
```diff
@@ -215,6 +312,14 @@ public function getResponse() {
 
+	/**
+	 * @param SS_HTTPResponse $response
+	 */
```
`@return $this`

tractorcow commented on control/Controller.php:
```diff
@@ -196,8 +289,12 @@ protected function handleAction($request, $action) {
 	}
 
+	/**
+	 * @param array $urlParams
```
`@return $this`

tractorcow commented on control/Controller.php:
```diff
@@ -83,30 +145,59 @@ public function Link() {
 	/**
+	 * {@inheritdoc}
+	 *
+	 * Also set the URLParams
```
`{@inheritdoc}` is probably not used properly here.

Please include `@param` and `@return` phpdoc.

assertchris replied:
If the intention is to add detail to the inherited phpdoc then it is used correctly.

dhensby replied:
<3

tractorcow commented on control/Controller.php:
```diff
@@ -76,6 +83,61 @@ public function init() {
+	 * This should be called on all controllers before handling requests
+	 */
+	public function callInit() {
+
```
No blank lines after the opening braces please. :D

tractorcow commented on control/Controller.php:
```diff
@@ -69,6 +69,13 @@ class Controller extends RequestHandler implements TemplateGlobalProvider {
 	 * @uses BasicAuth::requireLogin()
 	 */
 	public function init() {
```
How about making the method protected? :D

You can put the deprecation notice inside __call

dhensby replied:
But that is slow and nasty :(

I don't think we should start filling up __call with all our deprecation notices...

tractorcow replied:
Then just make it protected and NOT have the deprecation notice. ;)

tractorcow commented:
Like the PR... one functional suggestion given. :D

tractorcow commented on control/Controller.php:
```diff
@@ -76,6 +83,61 @@ public function init() {
+	 *
+	 * This should be called on all controllers before handling requests
+	 */
+	public function callInit() {
```
I would have gone with `doInit` to be consistent with the other `doMethods`, but that's just a point of artistic difference. :D

dhensby replied:
Good idea, I prefer that

dhensby commented:
Thinking about it, if this is going into master, then we don't need deprecation warnings... it's just pointless extra code; either this should go into 3.2 with deprecations or it should go in as a refactor to master..

What do you think @tractorcow?

dhensby commented:
OK, addressed most comments. Need thoughts on if this is 4.0 or 3.x

If it's 3.x then I'd be open to moving deprecation into `__call` otherwise I'd opt to just remove deprecation notices and put it in the upgrade notes...

tractorcow commented:
> If it's 3.x then I'd be open to moving deprecation into __call otherwise I'd opt to just remove deprecation notices and put it in the upgrade notes...

Sounds good. :D

dhensby commented:
Ok. Removed the depreciation handling which simplified things. 

We can still add a depreciation warning to 3.x that simply renames init to doInit to ease upgrade pain 

I just need to update the docs a bit more

dhensby commented:
Oh, and to add PRs to our other modules :P

tractorcow commented:
Sorry I forgot about this PR @dhensby ;_;

dhensby commented:
rebased

dhensby commented:
ping @tractorcow

assertchris commented:
I'm not a huge fan of init → doInit. Would prefer something like init → onInit.

Given the nature of this change (and especially since it is breaking), perhaps it would be good to get feedback from an additional developer? Perhaps @sminnee or @chillu, if they have some time?

dhensby commented:
It's only had 11 months for feedback ;)

The name of the function I think is neither here-nor-there. I can change it, but it is effort for a fairly subjective issue.

I do disagree with it being `onInit`, though. `on*` functions are usually event based (`onBeforeWrite`, `onAfterDelete`) where as this is the trigger so we are saying "do this thing" - ideally `init` should still be used but I chose not to so that the upgrade path was going to be far more simple this way (modules can keep their init functions and not needlessly rename them).

assertchris commented:
@dhensby 11 months is a long time, but it is still a breaking change and @tractorcow is the only developer to have given significant feedback on it yet. I made the `on*` comment because it was linked to from a thread, where @sminnee mentioned that as a preferred pattern. In that thread @tractorcow also uses `on*` as an example. It is subjective, sure.

> modules can keep their init functions and not needlessly rename them

Unless I'm mistaken, modules cannot keep their `init` methods because this PR changes the function visibility. Wouldn't they still have to change their `init` methods to `doInit`?

dhensby commented:
>> so any overloaded instances of this method need to have their visibility changed
> [Edited to replace some gubbins about init vs. doInit with a quote from the docs included in this PR]

I think it turns out that you can increase visibility without PHP throwing a shit-storm but not decrease, so that's actually bad docs by me... but people *should* change their visibility in maintained modules

dhensby commented:
Tests fail because the CMS test case is not pulling in these changes https://github.com/silverstripe/silverstripe-cms/pull/1169

tractorcow commented on control/Controller.php:
```diff
@@ -103,88 +129,129 @@ public function Link() {
 	public function handleRequest(SS_HTTPRequest $request, DataModel $model) {
-		if(!$request) user_error("Controller::handleRequest() not passed a request!", E_USER_ERROR);
+		if (!$request) {
+			user_error("Controller::handleRequest() not passed a request!", E_USER_ERROR);
```
I prefer InvalidArgumentException, but sure.

tractorcow commented on control/Controller.php:
```diff
@@ -103,88 +129,129 @@ public function Link() {
+	 */
+	protected function prepareResponse($response) {
+		if ($response instanceof SS_HTTPResponse) {
+			if (isset($_REQUEST['debug_request'])) {
```
TODO: Remove this debugging code.
